### DISCUSSION TITLE
How to encrypt a json object?

### DISCUSSION THREAD

**Discussion by DoneDeal0** [panva/jose#186] [Q&A] [answered]
Hi, 

I would like to store sensitive information in the local storage of my client. My idea is:
- retrieve info from the database, encrypt the object as a JOSE, send it to the client
- in the client, store the JOSE in local storage, and decipher it when an information is needed. 

Your library seems perfect for that. However, though I read your doc, I couldn't make it work. Here is what I got:

```js
app.get("/user", async (req, res) => {
  const encoder = new TextEncoder()
  const encryptedUser = await new CompactEncrypt(encoder.encode({name: "joe", age: 99, hobbies: ["swim", "eat"]}))
  .setProtectedHeader({ alg: 'RSA-OAEP-256', enc: 'A256GCM' })
  .encrypt("myPrivateKey")
  res.status(200).json({
    user: encryptedUser,
  });
});
```

Node throws this error: `invalid key input`. 

How to fix this, and which library would you recommend to decipher the JOSE token in the client? Is `jsonwebtoken` good enough if I pass it the right key?

Thanks for your help!

**Comment by panva** [comment, marked as answer]
First of all give #114 and the documentation for the encrypt method's key parameter type a read. See if it helps. If not, please come back.

**Reply by DoneDeal0** [reply]
Hi @panva, thanks for your help and the documentation links. It's much clearer indeed. So I've followed the doc, the JOSE is now correctly generated, but I can't decipher it so far. The reason is that the private key is generated inside the route, so the decoding operation is obviously doomed to fail. 

I can't generate it in the global scope, the generation will fail otherwise. Console logging the key displays `PrivateKeyObject { [Symbol(kKeyType)]: 'private' }` so it doesn't help either. I could send the privateKey in the response, but I'm afraid storing this key in localStorage would not be a smart option.

The best option would be to generate the private from a real password so I can store it in my frontend as well and perform the operation. Can it be done ?  If not, how would you proceed to solve this issue?

Thanks again for your time and your help, I highly appreciate it.

ps: here is the test code (I just tested the routes with postman, no frontend for the moment):

```js
const app = require("express")();
const cors = require("cors");
const { CompactEncrypt } = require('jose/jwe/compact/encrypt')
const { generateKeyPair } = require('jose/util/generate_key_pair')
const { compactDecrypt } = require('jose/jwe/compact/decrypt')

app.use(cors({ credentials: true, origin: ["http://localhost:3000"] }));

app.listen(8000, () => console.log(`server is listening on port 8000!`));

app.get("/user", async (req, res) => {
  const encoder = new TextEncoder()
  const { privateKey } =  await generateKeyPair('RSA-OAEP')
  const encryptedUser = await new CompactEncrypt(encoder.encode({name: "joe", age: 99, hobbies: ["swim", "eat"]}))
  .setProtectedHeader({ alg: 'RSA-OAEP-256', enc: 'A256GCM' })
  .encrypt(privateKey)
  res.status(200).json({
    user: encryptedUser,
  });
});

app.get("/decode", async (req, res) => {
  const decoder = new TextDecoder()
const { privateKey } =  await generateKeyPair('RSA-OAEP')
const { plaintext, protectedHeader } = await compactDecrypt(req.params.jwe, privateKey)
  res.status(200).json({
    user: decoder.decode(plaintext)
  });
});

```

**Reply by panva** [reply]
@DoneDeal0

1. while you are able to perform the encrypt operation with a KeyObject typed `private` what happens behind the scenes is that a public key is derived from your private input one. JOSE's asymmetric encryption is a public encryption, you encrypt for a recipient using its public key. The recipient then decrypts using the private one. This achieves confidentiality but it also means anyone with a knowledge of the public key is able to encrypt the data.
2. I would probably generate a non extractable keypair in the browser and send the public key to encrypt with back to the backend.
3. I would first sign the payload with a private key only known to your backend and then encrypt the resulting JWS as the plaintext of a JWE.
4. On the client side i would decrypt using the private key i've generated (and stored in Indexed DB) and then proceed to validate the signature of the decrypted JWS with the backends public key to know it was indeed my backend encrypting the data for me.

**Reply by DoneDeal0** [reply]
Ok, so in the end, I think it's easier to handle everything directly in the browser: 
- first call: `api.get("/user")`, receive json, encrypt it with pubic key, store encrypted user and private key in localStorage (I may use indexeddb later). 
- new call: check if user exists in local storage, if so, decipher it with private key, which is also stored in localStorage. 

So the encryption works fine, but when deciphering the user, I receive the error: 

> TypeError: invalid key input
    at decrypt (rsaes.js:21)
    at decryptKeyManagement (decrypt_key_management.js:60)
    at flattenedDecrypt (decrypt.js:95)
    at compactDecrypt (decrypt.js:15)

The code seems good to me:

```js
const getUser = async () => {
  try {
    if(localStorage.getItem("user")){
      const key = localStorage.getItem("key")
      const user = localStorage.getItem("user")
      const { plaintext, protectedHeader } = await compactDecrypt(user, key)
      return plaintext
    }
    const res = await api.get("/user");
    const { privateKey, publicKey } =  await generateKeyPair('RSA-OAEP')
    const encoder = new TextEncoder()
    const encryptedUser = await new CompactEncrypt(encoder.encode(res.data.user))
    .setProtectedHeader({ alg: 'RSA-OAEP-256', enc: 'A256GCM' })
    .encrypt(publicKey)
    localStorage.setItem("key", privateKey)
    localStorage.setItem("user", encryptedUser)
    return res;
  } catch (err) {
    return err;
  }
};
```
the private key looks like this:
```js
CryptoKey = {
    type: "private", 
    extractable: false, 
    algorithm: {
        hash: {name: "SHA-1"},
        modulusLength: 2048,
        name: "RSA-OAEP",
        publicExponent: [1, 0, 1],
    },
    extractable: false,
    type: "private",
    usages: ["decrypt", "unwrapKey"]
}
```
What goes wrong with the deciphering here?

**Reply by panva** [reply]
The alg on the key is wrong (RSA-OAEP is not the same as RSA-OAEP-256). But. If you're gonna do everything in the browser just use alg: dir and enc: A256GCM. 

This goes back to the jose algorithm selection guide. Public crypto is not fit for your case.

**Reply by panva** [reply]
Use the generate secret util to generate an A256GCM key.

**Reply by DoneDeal0** [reply]
Thank you so much panva! So I've followed your guidelines, but I still have the same error when deciphering: `TypeError: invalid key input`.  I generate a secret with `generateSecret('A256GCM')`, and encrypt the json with `{ alg: 'dir', enc: 'A256GCM' }`. 

When retrieving the key from localStorage, I receive `[object CryptoKey]`, so maybe it comes from here. 

The code is:

```js
const _getUserInfos = async () => {
    try {
      if(localStorage.getItem("user")){
        const key = await localStorage.getItem("key")
        const user = await localStorage.getItem("user")
        const { plaintext, protectedHeader } = await compactDecrypt(user, key)
        return plaintext
      }
      const res = await api.get("/user");
      const key =  await generateSecret('A256GCM')
      const encoder = new TextEncoder()
      const encryptedUser = await new CompactEncrypt(encoder.encode(res.data.user))
      .setProtectedHeader({ alg: 'dir', enc: 'A256GCM' })
      .encrypt(key)
      localStorage.setItem("key", key)
      localStorage.setItem("user", encryptedUser)
      return res;
    } catch (err) {
      return err;
    }
  };
```
I feel we're very close to solve this issue, but I don't see how to fix this so far. Is there something I'm doing wrong?

**Reply by panva** [reply]
i'm 100% confident you can't store a CryptoKey in localStorage, you have to use [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) to store the actual object instance. Orchestrating IndexedDB can be a pain, so i'd recommend a slim wrapper with a nice understandable API such as https://dexie.org/

<img width="576" alt="Screenshot 2021-04-21 at 17 41 24" src="https://user-images.githubusercontent.com/241506/115582013-d2e03500-a2c8-11eb-8d90-05271c651770.png">

See, it becomes a string when serialized like this.

**Reply by panva** [reply]
Furthermore `encoder.encode(res.data.user)` is not going to work most likely, you want to `JSON.stringify` it first.

**Reply by DoneDeal0** [reply]
Ok thanks! I need to learn indexedDB then. I'll let you know if I could solve the issue :)! Have a great evening!

**Reply by DoneDeal0** [reply]
Alright, I've learnt the basics, the secretKey is indeed properly stored in indexedDB. The encrypted json is now deciphered. However, the `plaintext` is an array of number (Uint8Array, 455 items), and not a proper object. 

I've tried to decode the Unit8Array with `new TextDecoder().decode(plaintext)`, but I've only got the encoded string.

It looks like this:

<img width="402" alt="Capture d’écran 2021-04-21 à 23 37 52" src="https://user-images.githubusercontent.com/43271780/115624230-9d543f80-a2fa-11eb-85ff-8e83cde90470.png">

I've tried with and without JSON.stringify in encoder.encode. The secret key looks ok though:
<img width="471" alt="Capture d’écran 2021-04-21 à 23 39 41" src="https://user-images.githubusercontent.com/43271780/115624421-e4dacb80-a2fa-11eb-9ab5-437f63646c18.png">

**Reply by panva** [reply]
Uint8array is the binary representation of your plaintext. First use the textdecoder, then json parse it.

**Reply by panva** [reply]
Its even shown in the decrypt functions doc examples.

**Reply by DoneDeal0** [reply]
Ok it works, all my apologies it was a stupid mistake of mine, it was midnight yesterday, I was definitely doing crap. I should have waited the morning with a fresh mind to come back at you. Have a great day, thanks again for your help!

**Reply by panva** [reply]
You're welcome, glad i could help. Please consider [supporting](https://github.com/panva/jose?sponsor=1) the library if it provides value to you or your company and this was of help to you. Supporting the library means, amongst other things, that such support will be available in the future.

**Comment by panva** [comment]
Second of all, this library is a universal/isomorphic javascript library. This means it works both in Node and the browser. The only difference is the KeyLike type that i've asked you to give a read in my previous comment. 

The jsonwebtoken is not fit for neither JWE nor browser runtime.
